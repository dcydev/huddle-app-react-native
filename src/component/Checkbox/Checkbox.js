import React, { useState, useEffect } from 'react';

import {
  Easing, Animated, View, TouchableWithoutFeedback,
} from 'react-native';

import Icon from '/component/Icon';
import Text from '/component/Text';

import styles, { disabledStyles, constants } from './Checkbox.styles';

const Checkbox = ({
  onChange,
  disabled,
  checked = false,
  style,
  round,
  label,
  hidden = false,
  testID = 'checkbox',
}) => {
  const [animateBackground] = useState(new Animated.Value(checked ? 1 : 0));
  const [animateCheck] = useState(new Animated.Value(checked ? 1 : 0));

  useEffect(() => {
    Animated.parallel([
      Animated.timing(animateBackground, {
        toValue: checked ? 1 : 0,
        useNativeDriver: true,
        duration: 250,
      }),
      // Start the check animation a bit after the background
      Animated.timing(animateCheck, {
        toValue: checked ? 1 : 0,
        useNativeDriver: true,
        duration: 150,
        delay: 100,
        easing: Easing.ease,
      }),
    ]).start();
  }, [checked]);

  const handlePress = () => {
    if (!disabled && onChange) {
      onChange(!checked);
    }
  };

  const backgroundStyle = [styles.background, disabled && disabledStyles.background];

  if (!disabled) {
    const borderRadius = round ? constants.size / 2 : constants.borderRadius;

    // If we are a square checkbox we will animate from a circle out to a square,
    // and the circle will start just on the inside of the border (that is this scale)
    // A round checkbox will just start as a smaller circle and animate to full
    const scale = round ? 0.7 : (constants.size - constants.borderWidth * 2) / constants.size;

    backgroundStyle.push({
      opacity: animateBackground,
      borderRadius: animateBackground.interpolate({
        inputRange: [0, 1],
        outputRange: [constants.size / 2, borderRadius],
      }),
      transform: [
        {
          scale: animateBackground.interpolate({
            inputRange: [0, 1],
            outputRange: [scale, 1],
          }),
        },
      ],
    });
  }

  // The checkbox animation is pretty subtle. We animate it up
  // and scale it slightly, and it goes from 0 -> 1 opacity
  // over the first 25% of the animation
  const checkStyle = [
    {
      opacity: animateCheck.interpolate({
        inputRange: [0, 0.25],
        outputRange: [0, 1],
      }),
      transform: [
        {
          translateY: animateCheck.interpolate({
            inputRange: [0, 1],
            outputRange: [2, 0],
          }),
        },
        {
          scale: animateCheck.interpolate({
            inputRange: [0, 1],
            outputRange: [0.85, 1],
          }),
        },
      ],
    },
  ];

  return (
    <TouchableWithoutFeedback
      onPress={handlePress}
      testID={testID}
    >
      <View style={[styles.wrapper, style]}>
        <View
          style={[
            styles.checkbox,
            disabled && disabledStyles.checkbox,
            round && styles.rounded,
            hidden && styles.placeholder,
          ]}
        >
          <Animated.View style={backgroundStyle} />
          <Icon style={checkStyle} name="checkmark" color="white" size={16} />
        </View>
        {!!label && (
          <Text style={styles.label} color={disabled ? 'medium' : 'dark'}>
            {label}
          </Text>
        )}
      </View>
    </TouchableWithoutFeedback>
  );
};

export default Checkbox;
