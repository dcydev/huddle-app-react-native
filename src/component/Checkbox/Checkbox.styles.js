import { StyleSheet } from 'react-native';

import globalStyles from '/styles';

const SIZE = 24;
const BORDER_RADIUS = 4;
const BORDER_WIDTH = 2;

export const constants = {
  size: SIZE,
  borderRadius: BORDER_RADIUS,
  borderWidth: BORDER_WIDTH,
};

export default StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: globalStyles.padding.sm,
  },
  checkbox: {
    width: SIZE,
    height: SIZE,
    borderRadius: BORDER_RADIUS,
    borderWidth: BORDER_WIDTH,
    borderColor: globalStyles.palette.grey02,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: globalStyles.palette.white,
  },
  rounded: {
    borderRadius: SIZE / 2,
  },
  background: {
    position: 'absolute',
    left: -BORDER_WIDTH,
    top: -BORDER_WIDTH,
    width: SIZE,
    height: SIZE,
    backgroundColor: globalStyles.palette.deepBlue,
  },
  label: {
    marginLeft: 60,
  },
  placeholder: {
    opacity: 0,
  },
});

export const disabledStyles = StyleSheet.create({
  checkbox: {
    backgroundColor: globalStyles.palette.grey03,
  },
  background: {
    opacity: 0,
  },
});
