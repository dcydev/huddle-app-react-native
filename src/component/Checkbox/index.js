export default from './Checkbox';
export { constants } from './Checkbox.styles';
