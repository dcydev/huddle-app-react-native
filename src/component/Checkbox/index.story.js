import React, { useState } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { storiesOf } from '@storybook/react-native';
import Container from 'storybook/Container';
import Checkbox from './Checkbox';

const DefaultCheckBox = () => {
  const [checked, setChecked] = useState(false);
  return (
    <Container>
      <Checkbox
        label="Sample Text"
        checked={checked}
        onChange={setChecked}
      />
    </Container>
  );
};

storiesOf('CheckBox', module).add('checkbox', () => (
  <DefaultCheckBox />
));
