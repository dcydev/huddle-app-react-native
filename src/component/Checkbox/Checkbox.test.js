import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react-native';
import Checkbox from './Checkbox';

describe('Checkbox', () => {
  let component;
  let props;
  let mockOnChange;

  const checkboxLabel = 'Mom, I\'m an engineer!';

  beforeEach(() => {
    mockOnChange = jest.fn();

    component = () => {
      const utils = render(
        <Checkbox
          onChange={mockOnChange}
          label={checkboxLabel}
          {...props}
        />,
      );

      return {
        ...utils,
      };
    };
  });

  afterEach(() => {
    jest.resetModules();
    cleanup();
  });

  describe('when checkbox is checked', () => {
    beforeAll(() => {
      props = {
        checked: true,
      };
    });

    it('renders correctly', () => {
      const { baseElement } = component();
      expect(baseElement).toMatchSnapshot();
    });

    it('shows correct label', () => {
      const { getByText } = component();
      expect(getByText(checkboxLabel)).toBeTruthy();
    });

    describe('when pressed', () => {
      beforeEach(() => {
        const { queryByTestId } = component();
        fireEvent.press(queryByTestId('checkbox'));
      });

      it('calls onChange prop', () => {
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).toBeCalledWith(false);
      });
    });

    describe('when checkbox is disabled', () => {
      beforeAll(() => {
        props = {
          checked: true,
          disabled: true,
        };
      });

      it('renders correctly', () => {
        const { baseElement } = component();
        expect(baseElement).toMatchSnapshot();
      });

      describe('when pressed', () => {
        beforeEach(() => {
          const { queryByTestId } = component();
          fireEvent.press(queryByTestId('checkbox'));
        });

        it('doesn\'t call onChange prop', () => {
          expect(mockOnChange).not.toBeCalled();
        });
      });
    });
  });

  describe('when checkbox is not checked', () => {
    beforeAll(() => {
      props = {
        checked: false,
      };
    });

    it('renders correctly', () => {
      const { baseElement } = component();
      expect(baseElement).toMatchSnapshot();
    });

    it('shows correct label', () => {
      const { getByText } = component();
      expect(getByText(checkboxLabel)).toBeTruthy();
    });

    describe('when pressed', () => {
      beforeEach(() => {
        const { queryByTestId } = component();
        fireEvent.press(queryByTestId('checkbox'));
      });

      it('calls onChange prop', () => {
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).toBeCalledWith(true);
      });
    });

    describe('when checkbox is disabled', () => {
      beforeAll(() => {
        props = {
          checked: false,
          disabled: true,
        };
      });

      it('renders correctly', () => {
        const { baseElement } = component();
        expect(baseElement).toMatchSnapshot();
      });

      it('doesn\'t call onChange prop', () => {
        const { queryByTestId } = component();
        fireEvent.press(queryByTestId('checkbox'));

        expect(mockOnChange).not.toBeCalled();
      });
    });
  });
});
