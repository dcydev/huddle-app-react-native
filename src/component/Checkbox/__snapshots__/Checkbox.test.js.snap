// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox when checkbox is checked renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableWithoutFeedback
      testID="checkbox"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "padding": 16,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderColor": "#d4d6db",
                "borderRadius": 4,
                "borderWidth": 2,
                "height": 24,
                "justifyContent": "center",
                "width": 24,
              },
              undefined,
              undefined,
              false,
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#203054",
                "borderRadius": 4,
                "height": 24,
                "left": -2,
                "opacity": 1,
                "position": "absolute",
                "top": -2,
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
                "width": 24,
              }
            }
          />
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 16,
                },
                Object {
                  "color": "white",
                  "opacity": 4,
                  "textShadowColor": "transparent",
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                },
                Object {
                  "fontFamily": "icomoon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          style={
            Array [
              undefined,
              Object {
                "color": "#203054",
                "fontFamily": "Lato",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 22,
              },
              false,
              Object {
                "marginLeft": 60,
              },
            ]
          }
        >
          Mom, I'm an engineer!
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;

exports[`Checkbox when checkbox is checked when checkbox is disabled renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableWithoutFeedback
      testID="checkbox"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "padding": 16,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderColor": "#d4d6db",
                "borderRadius": 4,
                "borderWidth": 2,
                "height": 24,
                "justifyContent": "center",
                "width": 24,
              },
              Object {
                "backgroundColor": "#e3e4e8",
              },
              undefined,
              false,
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#203054",
                "height": 24,
                "left": -2,
                "opacity": 0,
                "position": "absolute",
                "top": -2,
                "width": 24,
              }
            }
          />
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 16,
                },
                Object {
                  "color": "white",
                  "opacity": 4,
                  "textShadowColor": "transparent",
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                },
                Object {
                  "fontFamily": "icomoon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          style={
            Array [
              undefined,
              Object {
                "color": "#71798a",
                "fontFamily": "Lato",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 22,
              },
              false,
              Object {
                "marginLeft": 60,
              },
            ]
          }
        >
          Mom, I'm an engineer!
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;

exports[`Checkbox when checkbox is not checked renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableWithoutFeedback
      testID="checkbox"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "padding": 16,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderColor": "#d4d6db",
                "borderRadius": 4,
                "borderWidth": 2,
                "height": 24,
                "justifyContent": "center",
                "width": 24,
              },
              undefined,
              undefined,
              false,
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#203054",
                "borderRadius": 12,
                "height": 24,
                "left": -2,
                "opacity": 0,
                "position": "absolute",
                "top": -2,
                "transform": Array [
                  Object {
                    "scale": 0.8333333333333334,
                  },
                ],
                "width": 24,
              }
            }
          />
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 16,
                },
                Object {
                  "color": "white",
                  "opacity": 0,
                  "textShadowColor": "transparent",
                  "transform": Array [
                    Object {
                      "translateY": 2,
                    },
                    Object {
                      "scale": 0.85,
                    },
                  ],
                },
                Object {
                  "fontFamily": "icomoon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          style={
            Array [
              undefined,
              Object {
                "color": "#203054",
                "fontFamily": "Lato",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 22,
              },
              false,
              Object {
                "marginLeft": 60,
              },
            ]
          }
        >
          Mom, I'm an engineer!
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;

exports[`Checkbox when checkbox is not checked when checkbox is disabled renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableWithoutFeedback
      testID="checkbox"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "padding": 16,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderColor": "#d4d6db",
                "borderRadius": 4,
                "borderWidth": 2,
                "height": 24,
                "justifyContent": "center",
                "width": 24,
              },
              Object {
                "backgroundColor": "#e3e4e8",
              },
              undefined,
              false,
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#203054",
                "height": 24,
                "left": -2,
                "opacity": 0,
                "position": "absolute",
                "top": -2,
                "width": 24,
              }
            }
          />
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 16,
                },
                Object {
                  "color": "white",
                  "opacity": 0,
                  "textShadowColor": "transparent",
                  "transform": Array [
                    Object {
                      "translateY": 2,
                    },
                    Object {
                      "scale": 0.85,
                    },
                  ],
                },
                Object {
                  "fontFamily": "icomoon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          allowFontScaling={false}
          style={
            Array [
              undefined,
              Object {
                "color": "#71798a",
                "fontFamily": "Lato",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 22,
              },
              false,
              Object {
                "marginLeft": 60,
              },
            ]
          }
        >
          Mom, I'm an engineer!
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;
